File: src/main/java/cn/solarmoon/idyllic_food_diary/element/matter/cookware/kettle/KettleBlockEntity.java
package cn.solarmoon.idyllic_food_diary.element.matter.cookware.kettle;

import cn.solarmoon.idyllic_food_diary.feature.generic_recipe.water_boiling.IWaterBoilingRecipe;
import cn.solarmoon.idyllic_food_diary.feature.tea_brewing.IBrewingRecipe;
import cn.solarmoon.idyllic_food_diary.registry.common.IMBlockEntities;
import cn.solarmoon.solarmoon_core.api.tile.SyncedBlockEntity;
import cn.solarmoon.solarmoon_core.api.tile.fluid.TileTank;
import cn.solarmoon.solarmoon_core.api.tile.inventory.TileInventory;
import net.minecraft.core.BlockPos;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraftforge.fluids.capability.templates.FluidTank;
import net.minecraftforge.items.ItemStackHandler;

public class KettleBlockEntity extends SyncedBlockEntity implements IWaterBoilingRecipe, IBrewingRecipe {

    private final TileInventory inventory;
    private final TileTank fluidTank;

    private int boilTime;
    private int boilRecipeTime;
    private int brewTime;
    private int brewRecipeTime;

    public KettleBlockEntity(BlockPos pos, BlockState state) {
        super(IMBlockEntities.KETTLE.get(), pos, state);
        inventory = new TileInventory(3, 1, this);
        fluidTank = new TileTank(1000, this);
    }

    @Override
    public int getBoilRecipeTime() {
        return boilRecipeTime;
    }

    @Override
    public int getBoilTime() {
        return boilTime;
    }

    @Override
    public void setBoilRecipeTime(int time) {
        boilRecipeTime = time;
    }

    @Override
    public void setBoilTime(int time) {
        boilTime = time;
    }

    @Override
    public void setBrewTime(int time) {
        brewTime = time;
    }

    @Override
    public void setBrewRecipeTime(int time) {
        brewRecipeTime = time;
    }

    @Override
    public int getBrewTime() {
        return brewTime;
    }

    @Override
    public int getBrewRecipeTime() {
        return brewRecipeTime;
    }

    @Override
    public ItemStackHandler getInventory() {
        return inventory;
    }

    @Override
    public FluidTank getTank() {
        return fluidTank;
    }

}

File: src/main/kotlin/cn/solarmoon/idyllic_food_diary/element/matter/cookware/kettle/KettleBlockEntity.kt
package cn.solarmoon.idyllic_food_diary.element.matter.cookware.kettle

import cn.solarmoon.idyllic_food_diary.element.recipe.BrewingRecipe
import cn.solarmoon.idyllic_food_diary.element.recipe.WaterBoilingRecipe
import cn.solarmoon.idyllic_food_diary.registry.common.IFDBlockEntities
import cn.solarmoon.spark_core.api.blockentity.SyncedBlockEntity
import cn.solarmoon.spark_core.api.cap.fluid.FluidHandlerHelper
import cn.solarmoon.spark_core.api.cap.fluid.TileTank
import cn.solarmoon.spark_core.api.cap.item.ItemStackHandlerHelper
import cn.solarmoon.spark_core.api.cap.item.TileInventory
import cn.solarmoon.spark_core.api.recipe.processor.RecipeProcessorHelper
import cn.solarmoon.spark_core.registry.common.SparkDataComponents
import net.minecraft.core.BlockPos
import net.minecraft.core.HolderLookup
import net.minecraft.nbt.CompoundTag
import net.minecraft.world.level.block.state.BlockState
import net.neoforged.neoforge.fluids.SimpleFluidContent

class KettleBlockEntity(pos: BlockPos, state: BlockState) : SyncedBlockEntity(IFDBlockEntities.KETTLE.get(), pos, state) {

    val tank = TileTank(this, 1000)
    val inventory = TileInventory(this, 4, 1)
    val boil = WaterBoilingRecipe.Processor(this, tank)
    val brew = BrewingRecipe.Processor(this, inventory, tank)

    init {
        tank.fluid = components().getOrDefault(SparkDataComponents.SIMPLE_FLUID_CONTENT.value(), SimpleFluidContent.EMPTY).copy()
        RecipeProcessorHelper.createMap(this, boil, brew)
    }

    override fun saveAdditional(tag: CompoundTag, registries: HolderLookup.Provider) {
        super.saveAdditional(tag, registries)
        tag.put(FluidHandlerHelper.FLUID, tank.writeToNBT(registries, CompoundTag()))
        tag.put(ItemStackHandlerHelper.ITEM, inventory.serializeNBT(registries))
    }

    override fun loadAdditional(tag: CompoundTag, registries: HolderLookup.Provider) {
        super.loadAdditional(tag, registries)
        tank.readFromNBT(registries, tag.getCompound(FluidHandlerHelper.FLUID))
        inventory.deserializeNBT(registries, tag.getCompound(ItemStackHandlerHelper.ITEM))
    }

}